nomt/src/page_cache.rs

```diff
@@ -449,56 +560,63 @@ impl PageCache {
     ///
     /// This method is blocking, but doesn't suffer from the channel overhead.
     pub fn retrieve_sync(&self, page_id: PageId, hint_fresh: bool) -> Page {
-        let maybe_inflight = match self.shared.cached.entry(page_id.clone()) {
-            Entry::Occupied(mut o) => {
-                let page = o.get_mut();
-                match page {
-                    PageState::Cached(ref page) => {
-                        return Page {
-                            inner: page.clone(),
-                        };
-                    }
-                    PageState::Inflight(ref inflight) => {
-                        if hint_fresh {
-                            // pre-empt stale fetch.
-
-                            let page_data = Arc::new(PageData::pristine_empty(
-                                &self.shared.page_rw_pass_domain,
-                                page_id,
-                            ));
-                            let fresh_page = Page {
-                                inner: page_data.clone(),
-                            };
-
-                            inflight.complete_and_notify(fresh_page.clone());
-                            *page = PageState::Cached(page_data);
-                            return fresh_page;
-                        } else {
-                            Some(inflight.clone())
-                        }
-                    }
-                }
+        let shard_index = match self.shard_index_for(&page_id) {
+            None => {
+                let page_data = self.shared.root_page.read().clone();
+                return Page { inner: page_data };
             }
-            Entry::Vacant(v) => {
-                if hint_fresh {
-                    let page_data = Arc::new(PageData::pristine_empty(
-                        &self.shared.page_rw_pass_domain,
-                        page_id,
-                    ));
-                    let page = Page {
-                        inner: page_data.clone(),
-                    };
-                    v.insert(PageState::Cached(page_data));
-                    return page;
-                }
-                v.insert(PageState::Inflight(Arc::new(InflightFetch::new())));
-                None
+            Some(i) => i,
+        };
+
+        let mut shard = self.shard(shard_index).locked.lock();
+        if let Some(page) = shard.cached.get(&page_id) {
+            return Page {
+                inner: page.clone(),
+            };
+        }
+
+        let shard_mut = &mut *shard;
+        let existing_inflight = if let Some(inflight) = shard_mut.inflight.get(&page_id) {
```

Comment:
I think `existing_inflight` thing is not really elegant and a better solution would be to explicitly call `drop(shard_guard)` and then call wait within the if statement. You would have to clone out the inflight beforehand.

Reply:
can we punt on this? it was left over from before and we may refactor I/O more later

Reply:
ofc!